OpenOversight/app/commands.py

```diff
@@ -411,6 +421,388 @@ def bulk_add_officers(filename, no_create, update_by_name, update_static_fields)
+
+
+@click.command()
+@click.argument("department-name")
+@click.option("--officers-csv", type=click.Path(exists=True))
+@click.option("--assignments-csv", type=click.Path(exists=True))
+@click.option("--salaries-csv", type=click.Path(exists=True))
+@click.option("--links-csv", type=click.Path(exists=True))
+@click.option("--incidents-csv", type=click.Path(exists=True))
+@click.option("--force-create", is_flag=True, help="Only for development/testing!")
+@with_appcontext
+def advanced_csv_import(
+    department_name,
+    officers_csv,
+    assignments_csv,
+    salaries_csv,
+    links_csv,
+    incidents_csv,
+    force_create,
+):
+    """
+    Add or update officers, assignments, salaries, links and incidents from csv
+    files in the department DEPARTMENT_NAME.
+
+    The csv files are treated as the source of truth.
+    Existing entries might be overwritten as a result, backing up the
+    database and running the command locally first is highly recommended.
+
+    See the documentation before running the command.
+    """
+    if force_create and current_app.config["ENV"] == "production":
+        raise Exception("--force-create cannot be used in production!")
+    department_id = Department.query.filter_by(name=department_name).one().id
+
+    new_officers = {}
+    existing_officers = Officer.query.filter_by(department_id=department_id).all()
+    if force_create:
+        existing_officers = Officer.query.all()
+    id_to_officer = {officer.id: officer for officer in existing_officers}
+
+    counter = 0
+    if officers_csv is not None:
+        with open(officers_csv) as f:
+            csv_reader = csv.DictReader(f)
+            csv_reader.fieldnames = _unify_field_names(csv_reader.fieldnames)
+            _check_provided_fields(
+                csv_reader,
+                required_fields=["id", "department_name"],
+                optional_fields=[
+                    "last_name",
+                    "first_name",
+                    "middle_initial",
+                    "suffix",
+                    "race",
+                    "gender",
+                    "employment_date",
+                    "birth_year",
+                    "unique_internal_identifier",
+                    # the following are unused, but allowed since they are included in the csv output
+                    "badge_number",
+                    "job_title",
+                    "most_recent_salary",
+                    "last_employment_date",
+                    "last_employment_notice",
+                ],
+                csv_name="officers",
+            )
+
+            for row in csv_reader:
+                # can only update department with given name
+                assert row["department_name"] == department_name
+                row["department_id"] = department_id
+                connection_id = row["id"]
+                if row["id"].startswith("#"):
+                    row["id"] = ""
```

Comment:
this feels weird? Why just comment out the ID? Is this a dirty-data fix?

Reply:
This is my approach to be able to import foreign key relationships. Basically, if you set an id-field in the officer csv to `#1` and then use it in assignments csv in the field `officer_id``the assignment will be created for that officer. See the csv files in the tests for examples of how this looks like. I also tried to explain it in the added documentation.
Also, I am not sure if this is the best way to implement these kind of relations, so I am certainly open to suggestions to do it differently.